creator-node/src/fileManager.js

```diff
@@ -97,6 +97,217 @@ async function copyMultihashToFs(multihash, srcPath, logContext) {
+        },
+        logger,
+        logLabel: 'fetchFileFromTargetGatewayAndWriteToDisk',
+        options: {
+          retries: numRetries,
+          minTimeout: 3000
+        }
+      })
+
+      decisionTree.recordStage({
+        name: 'Found file from target gateway',
+        data: { targetGateway: contentUrl }
+      })
+
+      // If successful, will reach this point in code. Return out of fn
+      return
+    } catch (e) {
+      decisionTree.recordStage({
+        name: 'Error - Could not retrieve file from gateway',
+        data: {
+          url: contentUrl,
+          errorMsg: e.message,
+          multihash
+        }
+      })
+    }
+  }
+
+  // If file is not found in replica set, check network (remaining registered nodes)
+  try {
+    const found = await Utils.findCIDInNetwork(
+      path,
+      multihash,
+      logger,
+      libs,
+      /** trackId */ null,
+      /** excludeList */ targetGateways
+    )
+
+    if (!found) {
+      throw new Error(`Did not find multihash=${multihash} from network`)
+    }
+
+    decisionTree.recordStage({
+      name: 'Found file from network'
+    })
+
+    return
+  } catch (e) {
+    decisionTree.recordStage({
+      name: `Failed to find file from network`,
+      data: { errorMsg: e.message }
+    })
+  }
+
+  // error if file was not found on any gateway
+  const errorMsg = `Failed to retrieve file for multihash ${multihash} after trying entire network`
+  decisionTree.recordStage({
+    name: errorMsg
+  })
+  throw new Error(errorMsg)
+}
+
+/**
+ * Fetches a multihash via the /ipfs route, writes to disk, and verifies that the CID is what we
+ * expect it to be with retries
+ * @param {Object} param
+ * @param {function} param.bail the npm module async-retry bail fn
+ * @param {string} param.contentUrl the target content node ipfs gateway route
+ * @param {number} param.trackId the track id if one is associated with the multihash fetched
+ * @param {Object} param.decisionTree an instance of DecisionTree
+ * @param {string} param.path the path at which to save the fetched multihash from
+ * @param {Object} param.logger
+ */
+async function fetchFileFromTargetGatewayAndWriteToDisk({
+  bail,
+  contentUrl,
+  trackId,
+  multihash,
+  decisionTree,
+  path,
+  logger
+}) {
+  let response
+  try {
+    const fetchReqParams = {
+      method: 'get',
+      url: contentUrl,
+      responseType: 'stream',
+      timeout: 20000 /* 20 sec - higher timeout to allow enough time to fetch copy320 */
+    }
+
+    if (trackId) {
+      fetchReqParams.params = {
+        trackId
+      }
+    }
+
+    response = await axios(fetchReqParams)
+  } catch (e) {
+    if (
+      e.response?.status === 403 || // delist
+      e.response?.status === 401 || // unauth
+      e.response?.status === 400 // bad request
+    ) {
+      bail(
+        new Error(
+          `Content is delisted, request is unauthorized, or the request is bad with statusCode=${e.response?.status}`
+        )
```

Comment:
i don't think `bail` should be passed around to internal functions - confusing control flow here - `fetchFileFromTargetGatewayAndWriteToDisk()` as a pure function makes the request, throws error or returns response

the `asyncRetry()` inside `fetchFileFromNetworkAndWriteToDisk()` should handle this error inside its catch instead

wdyt?

Reply:
this is the one i feel strongly about btw

Reply:
yes I do think should have the asyncRetry fn handle bails. I just thought for the time being we can leave as is. do you have any suggestions about how we can change that?

Reply:
fwiw we do this already in TranscodeHandoffManager.js https://github.com/AudiusProject/audius-protocol/blob/master/creator-node/src/components/tracks/TrackTranscodeHandoffManager.js#L204

Reply:
i'm not suggesting we not use `bail()` here - that is def the right call, just that we not pass it into `fetchFileFromTargetGatewayAndWriteToDisk()` but rather handle the bail logic inside `fetchFileFromNetworkAndWriteToDisk` itself

the `asyncRetry()` wrapper is fine as is

Reply:
the transcodehandoffmanager does it correctly, sending you PR in a sec

Reply:
https://github.com/AudiusProject/audius-protocol/pull/3833

Reply:
nice